x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.http.retry;
+
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.inference.InferenceResults;
+import org.elasticsearch.xpack.inference.external.http.HttpResult;
+import org.elasticsearch.xpack.inference.external.http.sender.Sender;
+import org.elasticsearch.xpack.inference.logging.ThrottlerManager;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class RetryingHttpSender implements Retrier {
+    private static final Logger classLogger = LogManager.getLogger(RetryingHttpSender.class);
+
+    private final Sender sender;
+    private final ThrottlerManager throttlerManager;
+    private final Logger logger;
+    private final RetrySettings retrySettings;
+
+    public RetryingHttpSender(Sender sender, ThrottlerManager throttlerManager, Logger logger, RetrySettings retrySettings) {
+        this.sender = Objects.requireNonNull(sender);
+        this.throttlerManager = Objects.requireNonNull(throttlerManager);
+        this.logger = Objects.requireNonNull(logger);
+        this.retrySettings = Objects.requireNonNull(retrySettings);
+    }
+
+    @Override
+    public void send(HttpRequestBase request, ResponseHandler responseHandler, ActionListener<InferenceResults> listener) {
+        int failureCount = 0;
+        retryingSend(request, responseHandler, failureCount, listener);
+    }
+
+    private void retryingSend(
+        HttpRequestBase request,
+        ResponseHandler responseHandler,
+        int failureCount,
+        ActionListener<InferenceResults> listener
+    ) {
+        ActionListener<HttpResult> responseListener = ActionListener.wrap(result -> {
+            try {
+                responseHandler.validateResponse(throttlerManager, logger, request, result);
```

Comment:
Not all errors will be retryable. `ResponseHandler.validateResponse` could throw a `RetryableException` with the cause as a nested exception. I think that will make the handling logic easier and individual implementations of `ResponseHandler` can decide what is retryable.

Reply:
Good idea. I refactored this so it uses a `RetryException`. Let me know what you think. I'll add some more comments about some of the decisions I made. I have some general questions around whether the custom exception should be extending `ElasticsearchException` or using `Exception`.